aptos-move/block-executor/src/view.rs

```diff
@@ -50,37 +57,76 @@ pub(crate) enum ReadResult<V> {
 
 impl<
         'a,
-        K: ModulePath + PartialOrd + Ord + Send + Clone + Debug + Hash + Eq,
+        K: ModulePath + PartialOrd + Ord + Send + Sync + Clone + Debug + Hash + Eq,
         V: TransactionWrite + Send + Sync,
-    > MVHashMapView<'a, K, V>
+        X: Executable,
+    > MVHashMapView<'a, K, V, X>
 {
-    pub(crate) fn new(
-        versioned_map: &'a MVHashMap<K, V, ExecutableTestType>,
-        scheduler: &'a Scheduler,
-    ) -> Self {
+    pub(crate) fn new(versioned_map: &'a MVHashMap<K, V, X>, scheduler: &'a Scheduler) -> Self {
         Self {
             versioned_map,
             scheduler,
             captured_reads: RefCell::new(Vec::new()),
+            captured_executables: RefCell::new(Vec::new()),
         }
     }
 
     /// Drains the captured reads.
-    pub(crate) fn take_reads(&self) -> Vec<ReadDescriptor<K>> {
-        self.captured_reads.take()
+    pub(crate) fn take_captured_inputs(
+        self,
+    ) -> (Vec<(K, ReadDescriptor)>, Vec<(K, ExecutableDescriptor)>) {
+        (self.captured_reads.take(), self.captured_executables.take())
     }
 
-    // TODO: Actually fill in the logic to record fetched executables, etc.
+    /// Captures a key, executable descriptor pair if an executable is returned,
+    /// otherwise, may return a module blob without capturing anything (for the
+    /// VM to produce executable), or None, indicating the module blob must be
+    /// returned from storage. May internally wait for a dependency.
     fn fetch_code(
         &self,
         key: &K,
         txn_idx: TxnIndex,
-    ) -> anyhow::Result<MVCodeOutput<V, ExecutableTestType>, MVCodeError> {
+    ) -> Option<(ExecutableDescriptor, FetchedModule<X>)> {
+        use MVCodeError::*;
+        use MVCodeOutput::*;
+
+        loop {
+            match self.versioned_map.fetch_code(key, txn_idx) {
+                Ok(Executable((executable, descriptor))) => {
+                    self.captured_executables
```

Comment:
based on a discussion w. @sasha8, 
TODO: let's store a code dependency in both cases, both when an executable is returned and when a module is returned. This way, the validation logic will be correct even if some executable is only used for say verifying a different module and would not require the user of Executable trait to store and re-load the executable.

The drawback is that it removes an optimization where if by the time a thread stored an executable a different one was the right one, but this is likely not very common and more natural behavior to stick with the first module returned anyway and make it idempotent.